python/ray/serve/pipeline/generate.py

```diff
@@ -0,0 +1,170 @@
+        else:
+            init_kwargs[key] = value
+    return init_args, init_kwargs
+
+
+def transform_ray_dag_to_serve_dag(dag_node):
+    """
+    Transform a Ray DAG to a Serve DAG. Map ClassNode to DeploymentNode with
+    ray decorated body passed in, ans ClassMethodNode to DeploymentMethodNode.
+    """
+    if isinstance(dag_node, ClassNode):
+        deployment_name = _get_unique_deployment_name(dag_node)
+        ray_actor_options = _remove_non_default_ray_actor_options(
+            dag_node.get_options()
+        )
+        init_args, init_kwargs = _replace_init_args_with_deployment_handle(
+            dag_node.get_args(), dag_node.get_kwargs()
+        )
+        # Deployment class cannot bind with DeploymentNode
+        new_deployment = Deployment(
+            dag_node._body,
+            deployment_name,
+            DeploymentConfig(),
+            init_args=tuple(init_args),  # replace DeploymentNode with handle
+            init_kwargs=init_kwargs,
+            ray_actor_options=ray_actor_options,
+            _internal=True,
+        )
+        deployment_node = DeploymentNode(
+            new_deployment,
+            dag_node.get_args(),
+            dag_node.get_kwargs(),
+            # TODO: (jiaodong) Support .options(metadata=xxx) for deployment
+            {}, # Deployment options are not ray actor options.
+            other_args_to_resolve=dag_node.get_other_args_to_resolve(),
+        )
+        return deployment_node
+
+    elif isinstance(dag_node, ClassMethodNode):
+        other_args_to_resolve = dag_node.get_other_args_to_resolve()
+        # TODO: (jiaodong) Need to capture DAGNodes in the parent node
+        parent_deployment_node = other_args_to_resolve["parent_class_node"]
+
+        deployment_method_node = DeploymentMethodNode(
+            parent_deployment_node._body,
+            dag_node._method_name,
+            dag_node.get_args(),
+            dag_node.get_kwargs(),
+            dag_node.get_options(),
+            other_args_to_resolve=dag_node.get_other_args_to_resolve(),
+        )
+        return deployment_method_node
+    else:
+        # TODO: (jiaodong) Support FunctionNode
+        return dag_node
+
+
+def extract_deployments_from_serve_dag(
+    serve_dag_root: DAGNode,
+) -> List[Deployment]:
+    """Extract deployment python objects from a transformed serve DAG. Should
+    only be called after `transform_ray_dag_to_serve_dag`, otherwise nothing
+    to return.
+
+    Args:
+        serve_dag_root (DAGNode): Transformed serve dag root node.
+    Returns:
+        List[Deployment]: List of deployment python objects fetched from serve
+            dag.
+    """
+    deployments = {}
+
+    def extractor(dag_node):
+        if isinstance(dag_node, DeploymentNode):
+            deployment = dag_node._body
+            # In case same deployment is used in multiple DAGNodes
+            deployments[deployment.name] = deployment
+        # elif DeploymentMethodNode
```

Comment:
?

Reply:
I think there's a `DeploymentMethodNode` conditional for other DAG patterns (we had it as DeploymentMethodNode in demo) so just putting up a commented line here in case I forgot ..